### PR TITLE
Add I18n support for migrating existing apps

### PR DESCRIPTION
**Pull request by adammw**
:v:

/cc @zendesk/vegemite

### Description
Add I18n support for migrating existing apps.
Moves translation files to `src/translations/en.json` which are compiled into the generated js assets to be used by the I18n helpers. 

A webpack loader also makes a copy of the translations only containing the permitted 'app' keys (for app name, description, parameter translations in ZAM).

### References
* JIRA: https://zendesk.atlassian.net/browse/AF-355

### CODE REVIEW

File: src/javascripts/i18n.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+}
+
+module.exports = Object.freeze({
+  t: function(app, key, context) {
+    var keyType = typeof key;
+    if (keyType !== 'string') {
+      throw new Error(`Translation key must be a string, got: ${keyType}`);
+    }
+    var template = _.at(app.translations, key)[0];
```
Comment (sandlerr) [inline src/javascripts/i18n.js]: don't need to require _?

**Comment by sandlerr** [comment]
is there a way to not use the Handlebars compiler at runtime?

File: src/javascripts/i18n.js (hunk 1 above)
Comment (adammw) [inline src/javascripts/i18n.js]: lodash is included at runtime by jsdelivr rather than by webpack, so no require...

**Comment by danielbreves** [comment]
Nice work! 👍

Could we move BaseApp and I18n to a lib directory? I feel like having all these extra files may confuse beginners, so it would nice to hide them from sight somewhere. Same with loaders, but not sure what to do there, maybe we can extract them all to a npm module at some point.

We'll need document this on the README.

**Comment by adammw** [comment]
💇  @sandlerr 

@danielbreves, the readme here doesn't seem to go into detail about how all the parts of the scaffold work, just how to set it up and build it. do we document it elsewhere?

File: loaders/translations-loader.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+  });
+  return flattened;
+}
+
+function compileTranslations(translations) {
+  var flattenedTranslations = flatten(translations);
+  var rows = Object.keys(flattenedTranslations).map(function(translationKey) {
+    var template = hb.precompile(flattenedTranslations[translationKey]);
+    return `${JSON.stringify(translationKey)}: (Handlebars["default"] || Handlebars).template(${template})`;
```
Comment (sandlerr) [inline loaders/translations-loader.js]: could use `Handlebars.defaut`

**Comment by sandlerr** [comment]
:+1:

File: loaders/translations-loader.js (hunk 2 above)
Comment (adammw) [inline loaders/translations-loader.js]: Hmm, yeah didn't notice that. Tbh, i'm not 100% sure how this code works, it's based off: https://github.com/pcardune/handlebars-loader/blob/master/index.js#L287-L289

File: loaders/translations-loader.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+var loaderUtils = require('loader-utils');
```
Comment (danielbreves) [inline loaders/translations-loader.js]: Can we move /loaders to a /lib folder as well? That way we can document that everything under /lib and /src/lib are part of the scaffold tooling and don't need to be modified.

**Comment by adammw** [comment]
💇  @danielbreves

File: lib/loaders/translations-loader.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  var translationsOutput = {
+    _warning: `AUTOMATICALLY GENERATED FROM ${jsonPath} - DO NOT MODIFY THIS FILE DIRECTLY`,
+    app: {}
+  };
+  permittedKeys.forEach(function(key) {
+    if (key in translations.app) {
+      translationsOutput.app[key] = translations.app[key];
+    }
+  });
+  return JSON.stringify(translationsOutput, 2, 4);
```
Comment (danielbreves) [inline lib/loaders/translations-loader.js]: Can we name the magic numbers?

**Comment by danielbreves** [comment]
👍

**Pull request merged**